### PR TITLE
Allow early exit from puzzle map during reveal animation

### PR DESCRIPTION
**Pull request by Kazhuu**
Solves https://github.com/ihhub/fheroes2/issues/7448 by allowing exiting puzzle map during reveal animation.

### CODE REVIEW

**Comment by Kazhuu** [comment]
It seems Vita build is broken at the moment?

**Comment by oleg-derevenetz** [comment]
> It seems Vita build is broken at the moment?

Please see #7486.

File: src/fheroes2/kingdom/puzzle.cpp [hunk 1]
```diff
@@ -103,7 +103,8 @@ namespace
         }
     }
 
-    void revealPuzzle( const Puzzle & pzl, const fheroes2::Image & sf, int32_t dstx, int32_t dsty, const std::function<fheroes2::Rect()> * drawControlPanel = nullptr )
+    void revealPuzzle( const Puzzle & pzl, const fheroes2::Image & sf, int32_t dstx, int32_t dsty, fheroes2::Button & buttonExit,
```
Comment (ihhub) [inline src/fheroes2/kingdom/puzzle.cpp:106]: Would it be better to just return true or false from this function and then check the return value from the place where we call the function? Passing information from a caller of a function to the function must be the last resort.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Kazhuu , I left a suggestion regarding the proposed approach. Would you mind to take a look?

**Comment by Kazhuu** [comment]
> Hi @Kazhuu , I left a suggestion regarding the proposed approach. Would you mind to take a look?

Okay I returned the first implementation with boolean return variable. Can you @ihhub take a look?

File: src/fheroes2/kingdom/puzzle.cpp [hunk 2]
```diff
@@ -146,6 +152,7 @@ namespace
                 assert( alpha >= 0 );
             }
         }
+        return false;
```
Comment (ihhub) [inline src/fheroes2/kingdom/puzzle.cpp:155]: I recommend to add an empty line before this return to have a visual separation of the code parts.

**Comment by Kazhuu** [comment]
Hello @ihhub I was thinking because this version also need to pass exit button as a parameter, like you mentioned is not good. But do you think actually returning boolean value adds more code on the caller side and makes the code more complicated? Would you reconsider the implementation where the `revealPuzzle` function takes care of the exit button handling. Then I can remove those two while loops from the caller side. What do you think?

**Comment by Kazhuu** [comment]
Hello @ihhub I made the changes you requested. Can you have a look again, thanks!

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@Kazhuu , big thanks for this change!

**Pull request merged**